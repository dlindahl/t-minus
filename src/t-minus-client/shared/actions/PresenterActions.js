import ClockActions from '../actionTypes/ClockActionTypes'
import emptyObj from 'empty/object'
import { omit } from 'lodash'
import PresenterActions from '../actionTypes/PresenterActionTypes'

function parseEventData (data = '{}') {
  try {
    data = JSON.parse(data)
  } catch (error) {
    data = {
      payload: error,
      type: ClockActions.CLOCK_DATA_ERROR
    }
  }
  return data
}

/*
 Initializes the Presenter View's app state so that its in sync with the
 Client Window. Omits the `meta` reducer to avoid circular references.
*/
function initPresenterState (appWindow, store) {
  const state = store.getState()
  state.meta.presenterWin.postMessage(JSON.stringify({
    payload: omit(state, 'meta'),
    type: PresenterActions.PRESENTER_STATE_CHANGED
  }), appWindow.location.origin)
}

export function handlePresenterMessage (appWindow, store, event) {
  if (event.origin !== appWindow.location.origin) {
    return emptyObj
  }
  const data = parseEventData(event.data)
  switch (data.type) {
    case PresenterActions.PRESENTER_WINDOW_READY:
      initPresenterState(appWindow, store)
      return emptyObj
    default:
      // Dispatch all other postMessage events into the Redux store
      return store.dispatch(data)
  }
}

/*
 Tells the Client Window that the Presenter app has mounted and is ready to
 receive state and messages
*/
export function presenterReady (presenterWin) {
  if (!presenterWin.opener) {
    console.warn('Client Window not found, cannot notify Presenter is ready!')
    return {
      payload: emptyObj,
      type: PresenterActions.CLIENT_WINDOW_NOT_FOUND
    }
  }
  const appWindow = presenterWin.opener
  const msg = {
    payload: emptyObj,
    type: PresenterActions.PRESENTER_WINDOW_READY
  }
  presenterWin.addEventListener('beforeunload', function onPresenterUnload (event) {
    appWindow.postMessage(JSON.stringify({
      payload: emptyObj,
      type: PresenterActions.PRESENTER_WINDOW_CLOSED
    }), presenterWin.location.origin)
    presenterWin.removeEventListener('beforeunload', onPresenterUnload)
  })
  appWindow.postMessage(JSON.stringify(msg), presenterWin.location.origin)
  // Dispatch an event that can keep track of the Client Window object
  return {
    payload: {
      clientWin: presenterWin.opener
    },
    type: PresenterActions.PRESENTER_WINDOW_READY
  }
}

// Spawns a new window that contains the "Presenter" view of the clock display
export function spawnPresenterWindow (clientWin = window) {
  return (dispatch, getState) => {
    const { meta } = getState()
    if (meta.presenterWin) {
      console.warn('Presenter Window already spawned. If closed, it was not cleaned up properly.')
      return
    }
    const presenterWin = clientWin.open('presenter.html')
    dispatch({
      payload: { presenterWin },
      type: PresenterActions.PRESENTER_WINDOW_SPAWNED
    })
    clientWin.addEventListener('beforeunload', function handleClientClose () {
      if (!presenterWin.closed) {
        presenterWin.close()
      }
    })
  }
}
