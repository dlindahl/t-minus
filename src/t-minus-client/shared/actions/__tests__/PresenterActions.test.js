import * as presenter from '../PresenterActions';
import configureStore from 'redux-mock-store';
import emptyObj from 'empty/object';
import mute from 'mute';
import { omit } from 'lodash';

const mockStore = configureStore();
const mockWindow = () => ({
  addEventListener: jest.fn(),
  close: jest.fn(),
  location: {
    origin: 'ORIGIN'
  },
  open: jest.genMockFunction(),
  postMessage: jest.fn(),
  removeEventListener: jest.fn()
});
const DATA = {
  default: {
    type: 'TEST_TYPE',
    payload: {
      foo: 'bar'
    }
  },
  ready: {
    type: 'PRESENTER_WINDOW_READY',
    payload: {
      foo: 'baz'
    }
  }
};
const STATE = {
  default: () => ({
    foo: {
      bar: 'baz'
    },
    meta: {
      presenterWin: mockWindow()
    }
  }),
  noPresenter: {
    meta: emptyObj
  }
};

describe('PresenterActions', () => {
  describe('#handlePresenterMessage', () => {
    it('relays the data to the redux store', () => {
      const appWindow = mockWindow();
      const store = mockStore();
      const event = {
        data: JSON.stringify(DATA.default),
        origin: appWindow.location.origin
      };
      presenter.handlePresenterMessage(appWindow, store, event);
      const actions = store.getActions();
      expect(actions.length).toBe(1);
      expect(actions[0]).toMatchSnapshot();
    });

    it('relays an error to the redux store when the data is not parseable JSON', () => {
      const appWindow = mockWindow();
      const event = {
        data: '{{[[',
        origin: appWindow.location.origin
      };
      const store = mockStore();
      presenter.handlePresenterMessage(appWindow, store, event);
      const actions = store.getActions();
      expect(actions.length).toBe(1);
      expect(actions[0].type).toBe('CLOCK_DATA_ERROR');
      expect(actions[0].payload).toMatchSnapshot();
    });

    it('does nothing when the origin does not match', () => {
      const appWindow = mockWindow();
      const store = mockStore();
      const event = {
        data: JSON.stringify(emptyObj),
        origin: 'MISMATCH'
      };
      presenter.handlePresenterMessage(appWindow, store, event);
      const actions = store.getActions();
      expect(actions.length).toBe(0);
    });

    it('initializes the presenter window state when the window is ready', () => {
      const appWindow = mockWindow();
      const store = mockStore(STATE.default());
      const event = {
        data: JSON.stringify(DATA.ready),
        origin: appWindow.location.origin
      };
      presenter.handlePresenterMessage(appWindow, store, event);
      const mockPostMsg = store.getState().meta.presenterWin.postMessage;
      expect(mockPostMsg).toHaveBeenCalled();
      expect(mockPostMsg.mock.calls).toMatchSnapshot();
    });
  });

  describe('#presenterReady', () => {
    it('listens for the presenter window closing', () => {
      const store = mockStore();
      const appWindow = mockWindow();
      const presenterWin = mockWindow();
      presenterWin.opener = appWindow;
      presenter.presenterReady(presenterWin);
      const unloadEvent = presenterWin.addEventListener.mock.calls[0];
      unloadEvent[1]();
      expect(appWindow.postMessage).toHaveBeenCalledTimes(2);
      expect(appWindow.postMessage.mock.calls[1]).toMatchSnapshot();
    });

    it('tells the app window that the presenter window is ready', () => {
      const store = mockStore();
      const appWindow = mockWindow();
      const presenterWin = mockWindow();
      presenterWin.opener = appWindow;
      const payload = presenter.presenterReady(presenterWin);
      expect(payload).toMatchSnapshot();
      expect(appWindow.postMessage).toHaveBeenCalled();
      expect(appWindow.postMessage.mock.calls).toMatchSnapshot();
    });

    it('dispatches a NOT_FOUND event when the app window cannot be found', () => {
      const presenterWin = mockWindow();
      console.warn = jest.genMockFunction();
      const payload = presenter.presenterReady(presenterWin);
      expect(console.warn).toHaveBeenCalledTimes(1);
      expect(payload).toMatchSnapshot();
    });
  });

  describe('#spawnPresenterWindow', () => {
    it('opens a new presenter window', () => {
      const store = mockStore(STATE.noPresenter);
      const clientWin = mockWindow();
      presenter.spawnPresenterWindow(clientWin)(store.dispatch, store.getState);
      expect(clientWin.open).toHaveBeenCalledWith('presenter.html');
    });

    it('dispatches a PRESENTER_WINDOW_SPAWNED event when the presenter window has spawned', () => {
      const store = mockStore(STATE.noPresenter);
      const clientWin = mockWindow();
      const presenterWin = mockWindow();
      clientWin.open.mockReturnValueOnce(presenterWin);
      presenter.spawnPresenterWindow(clientWin)(store.dispatch, store.getState);
      const actions = store.getActions();
      expect(actions.length).toBe(1);
      expect(actions[0]).toMatchSnapshot();
    });

    it('closes the presenter window when the client window is closed', () => {
      const store = mockStore(STATE.noPresenter);
      const clientWin = mockWindow();
      const presenterWin = mockWindow();
      clientWin.open.mockReturnValueOnce(presenterWin);
      presenter.spawnPresenterWindow(clientWin)(store.dispatch, store.getState);
      const handler = clientWin.addEventListener.mock.calls[0];
      handler[1]()
      expect(presenterWin.close).toHaveBeenCalled();
    });

    it('does nothing when the presenter window already exists', () => {
      const store = mockStore(STATE.default);
      console.warn = jest.genMockFunction();
      presenter.spawnPresenterWindow()(store.dispatch, store.getState);
      expect(console.warn).toHaveBeenCalledTimes(1);
    });
  });
});
