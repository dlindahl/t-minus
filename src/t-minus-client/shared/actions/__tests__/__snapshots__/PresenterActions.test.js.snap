exports[`PresenterActions #handlePresenterMessage initializes the presenter window state when the window is ready 1`] = `
Array [
  Array [
    "{\"type\":\"PRESENTER_STATE_CHANGED\",\"payload\":{\"foo\":{\"bar\":\"baz\"}}}",
    "ORIGIN",
  ],
]
`;

exports[`PresenterActions #handlePresenterMessage relays an error to the redux store when the data is not parseable JSON 1`] = `[SyntaxError: Unexpected token { in JSON at position 1]`;

exports[`PresenterActions #handlePresenterMessage relays the data to the redux store 1`] = `
Object {
  "payload": Object {
    "foo": "bar",
  },
  "type": "TEST_TYPE",
}
`;

exports[`PresenterActions #presenterReady dispatches a NOT_FOUND event when the app window cannot be found 1`] = `
Object {
  "payload": Object {},
  "type": "CLIENT_WINDOW_NOT_FOUND",
}
`;

exports[`PresenterActions #presenterReady listens for the presenter window closing 1`] = `
Array [
  "{\"type\":\"PRESENTER_WINDOW_CLOSED\",\"payload\":{}}",
  "ORIGIN",
]
`;

exports[`PresenterActions #presenterReady tells the app window that the presenter window is ready 1`] = `
Object {
  "payload": Object {
    "clientWin": Object {
      "addEventListener": [Function],
      "close": [Function],
      "location": Object {
        "origin": "ORIGIN",
      },
      "open": [Function],
      "postMessage": [Function],
      "removeEventListener": [Function],
    },
  },
  "type": "PRESENTER_WINDOW_READY",
}
`;

exports[`PresenterActions #presenterReady tells the app window that the presenter window is ready 2`] = `
Array [
  Array [
    "{\"type\":\"PRESENTER_WINDOW_READY\",\"payload\":{}}",
    "ORIGIN",
  ],
]
`;

exports[`PresenterActions #spawnPresenterWindow dispatches a PRESENTER_WINDOW_SPAWNED event when the presenter window has spawned 1`] = `
Object {
  "payload": Object {
    "presenterWin": Object {
      "addEventListener": [Function],
      "close": [Function],
      "location": Object {
        "origin": "ORIGIN",
      },
      "open": [Function],
      "postMessage": [Function],
      "removeEventListener": [Function],
    },
  },
  "type": "PRESENTER_WINDOW_SPAWNED",
}
`;
